pkg/odo/genericclioptions/context_test.go

```diff
@@ -0,0 +1,424 @@
+				},
+			},
+			expectedErr: "",
+			expected: Context{
+				internalCxt: internalCxt{
+					project:          "myproject",
+					application:      "myapp",
+					component:        "mycomponent",
+					componentContext: filepath.Join(prefixDir, "myapp"),
+					outputFlag:       "",
+					devfilePath:      filepath.Join(prefixDir, "myapp", "devfile.yaml"),
+				},
+			},
+		},
+		{
+			name: "flags set, needDevfile, .devfile.yaml is present",
+			input: input{
+				needDevfile:   true,
+				isOffline:     true,
+				workingDir:    filepath.Join(prefixDir, "myapp"),
+				projectFlag:   "myproject",
+				appFlag:       "myapp",
+				componentFlag: "mycomponent",
+				outputFlag:    "",
+				allFlagSet:    false,
+				populateWorkingDir: func(fs filesystem.Filesystem) {
+					_ = fs.MkdirAll(filepath.Join(prefixDir, "myapp", ".odo", "env"), 0755)
+					_ = fs.WriteFile(filepath.Join(prefixDir, "myapp", ".odo", "env", "env.yaml"), []byte{}, 0644)
+					_ = fs.WriteFile(filepath.Join(prefixDir, "myapp", ".devfile.yaml"), []byte(devfileYAML), 0644)
+				},
+			},
+			expectedErr: "",
+			expected: Context{
+				internalCxt: internalCxt{
+					project:          "myproject",
+					application:      "myapp",
+					component:        "mycomponent",
+					componentContext: filepath.Join(prefixDir, "myapp"),
+					outputFlag:       "",
+					devfilePath:      filepath.Join(prefixDir, "myapp", ".devfile.yaml"),
+				},
+			},
+		},
+		{
+			name: "flags set, needDevfile, devfile.yaml is present",
+			input: input{
+				needDevfile:   true,
+				isOffline:     true,
+				workingDir:    filepath.Join(prefixDir, "myapp"),
+				projectFlag:   "myproject",
+				appFlag:       "myapp",
+				componentFlag: "mycomponent",
+				outputFlag:    "",
+				allFlagSet:    false,
+				populateWorkingDir: func(fs filesystem.Filesystem) {
+					_ = fs.MkdirAll(filepath.Join(prefixDir, "myapp", ".odo", "env"), 0755)
+					_ = fs.WriteFile(filepath.Join(prefixDir, "myapp", ".odo", "env", "env.yaml"), []byte{}, 0644)
+					_ = fs.WriteFile(filepath.Join(prefixDir, "myapp", "devfile.yaml"), []byte(devfileYAML), 0644)
+				},
+			},
+			expectedErr: "",
+			expected: Context{
+				internalCxt: internalCxt{
+					project:          "myproject",
+					application:      "myapp",
+					component:        "mycomponent",
+					componentContext: filepath.Join(prefixDir, "myapp"),
+					outputFlag:       "",
+					devfilePath:      filepath.Join(prefixDir, "myapp", "devfile.yaml"),
+				},
+			},
+		},
+		{
+			name: "no env file",
+			input: input{
+				needDevfile: false,
+				isOffline:   true,
+				workingDir:  filepath.Join(prefixDir, "myapp"),
+				populateWorkingDir: func(fs filesystem.Filesystem) {
+				},
+			},
+			expectedErr: "the current directory does not represent an odo component",
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+
+			os.RemoveAll(prefixDir)
+			defer os.RemoveAll(prefixDir)
```

Comment:
Do we need to RemoveAll and defer back to it? Shouldn't using `defer` suffice?

Reply:
The first one is to cleanup before execution, in case any other process terminated without cleaning it up.
The second is to clean it up after execution